standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/StatisticsTestUtils.java

```diff
@@ -0,0 +1,788 @@
+    colStats.setEngine(HIVE_ENGINE);
+    return colStats;
+  }
+
+  /**
+   * Creates an FM sketch object initialized with the given values.
+   * @param values the values to be added
+   * @return an FM sketch initialized with the given values.
+   */
+  public static FMSketch createFMSketch(long... values) {
+    FMSketch fm = new FMSketch(1);
+    for (long value : values) {
+      fm.addToEstimator(value);
+    }
+    return fm;
+  }
+
+  /**
+   * Creates an FM sketch object initialized with the given values.
+   * @param values the values to be added
+   * @return an FM sketch initialized with the given values.
+   */
+  public static FMSketch createFMSketch(String... values) {
+    FMSketch fm = new FMSketch(1);
+    for (String value : values) {
+      fm.addToEstimator(value);
+    }
+    return fm;
+  }
+
+  /**
+   * Creates an HLL object initialized with the given values.
+   * @param values the values to be added
+   * @return an HLL object initialized with the given values.
+   */
+  public static HyperLogLog createHll(long ... values) {
+    HyperLogLog hll = HyperLogLog.builder().build();
+    for (long value : values) {
+      hll.addLong(value);
+    }
+    return hll;
+  }
+
+  /**
+   * Creates an HLL object initialized with the given values.
+   * @param values the values to be added
+   * @return an HLL object initialized with the given values.
+   */
+  public static HyperLogLog createHll(String ... values) {
+    HyperLogLog hll = HyperLogLog.builder().build();
+    for (String value : values) {
+      hll.addBytes(value.getBytes());
+    }
+    return hll;
+  }
+
+  private static void assertBitVector(Object hll, boolean isSetBitVectors, byte[] bitVector) {
+    if (hll == null) {
+      Assert.assertFalse("bitVector should not be set", isSetBitVectors);
+      Assert.assertNull("bitVector should be non null", bitVector);
+    } else {
+      Assert.assertTrue("bitVector should be set", isSetBitVectors);
+      byte[] bytes;
+      if (hll instanceof HyperLogLog) {
+        bytes = ((HyperLogLog) hll).serialize();
+      } else if (hll instanceof FMSketch) {
+        bytes = ((FMSketch) hll).serialize();
+      } else {
+        throw new IllegalArgumentException("Unsupported HLL class: " + hll.getClass().getName());
+      }
+      // hll are left as-is
+      Assert.assertArrayEquals("hll mismatch", bytes, bitVector);
+    }
+  }
+
+  /*------------ LONG-RELATED METHODS ------------*/
+  /**
+   * Method for creating statistics for a column of the Long type family.
+   * @param numNulls the number of null values
+   * @param numDVs the number of distinct values
+   * @param low the low value
+   * @param high the high value
+   * @param hll the HLL value
+   * @return a statistics data object for a column of the Long type family with the sought values.
+   */
+  public static ColumnStatisticsData createLongStats(long numNulls, long numDVs, Long low, Long high, Object hll) {
```

Comment:
There are a lot of `createXxStats` utility methods here which are bit repetitive (code fragments resemble a lot). They also force the caller to pass a value for every parameter and from what I can see from the tests we don't always need to set everything. Also the code around hll is not type safe and requires the programmer to look into the method in order to understand what are the valid values.

Have you thought about using a builder pattern for creating the stats? 

```java
    new ColStatsBuilder().nulls(1).dvs(2).low(1).high(4).hll(1, 3).buildLong();
```

With the builder pattern we may be able to avoid the aforementioned disadvantages. Using a bit of reflection could also allow us to cut down the size of this class.

Reply:
I commited a prototype with the builder idea here: https://github.com/apache/hive/commit/09fef05909d1b2950681b62d8f50899ce9fae148

Reply:
I have re-used the builder because it's a great idea, the only "real" change I did to it is to have `lowX()`/`highX()` methods for each of the data types supporting low/high statistics, because those two methods are (the only) depending on the underlying data type, and as you suggested in another part, it's better to have type safety and not use `Object` as a param type, whenever possible.

Reply:
I tried to group the lowX/highX methods together using generics (check https://github.com/apache/hive/pull/3339/commits/c9fa7b658ac39676bb2418cbe99b735507de5d49 
https://github.com/apache/hive/pull/3339/commits/43aba5b9f76a56febcafe9513f035085749fddfa). WDYT?

Reply:
I also pushed some more changes around the builder check them out and let me know how it looks.

Reply:
It looks great, I like the idea of making it generic!